### PR TITLE
Optimize recode for the large number of categories when the categories to be recoded are specified as arrays

### PR DESCRIPTION
**Pull request by pgagarinov**
Should solve #343

### CODE REVIEW

File: src/recode.jl [hunk 1]
```diff
@@ -152,7 +170,7 @@ function recode!(dest::CategoricalArray{T}, src::AbstractArray, default::Any, pa
         # isless may throw an error, e.g. for AbstractArray{T} of unordered T
         try
             sort!(oldlevels)
-        catch e
+    catch e
```
Comment (nalimilan) [inline src/recode.jl:173]:
```suggestion
        catch e
```

File: src/recode.jl [hunk 2]
```diff
@@ -161,8 +179,8 @@ function recode!(dest::CategoricalArray{T}, src::AbstractArray, default::Any, pa
     dest
 end
 
-function recode!(dest::CategoricalArray{T, N, R}, src::CategoricalArray,
-                 default::Any, pairs::Pair...) where {T, N, R<:Integer}
+function recode!(dest::CategoricalArray{T,N,R}, src::CategoricalArray,
```
Comment (nalimilan) [inline src/recode.jl:182]: Please revert all these unrelated spacing changes.

File: test/16_recode.jl [hunk 3]
```diff
@@ -1,14 +1,41 @@
...
+using CategoricalArrays:DefaultRefType
 
 if VERSION < v"0.7.0-"
-    using CategoricalArrays: replace!
+    using CategoricalArrays:replace!
 end
 
 const ≅ = isequal
 
+function test_recode_performance(orig_vec, cat2merge_vec)
```
Comment (nalimilan) [inline test/16_recode.jl:12]: This would be more appropriate in the benchmarks folder (see how to use the PkgBenchmark/BenchmarkTools syntax there).

File: src/recode.jl [hunk 4]
```diff
@@ -49,16 +49,32 @@ A user defined type could override this method to define an appropriate test fun
...
+    pair_vec = Vector{Pair}(undef, length(pairs))
+    for (ind_pair, p) in enumerate(pairs)
+        if typeof(p.first) <: AbstractArray
+            pair_vec[ind_pair] = Set(p.first) => p.second
+        else
+            pair_vec[ind_pair] = p
+        end
+    end
+    return pair_vec
+end
```
Comment (nalimilan) [inline src/recode.jl:63]: It would be more logical and possibly more efficient to only define an `optimize_pair` function operating on a single value at a time, and do `opt_pairs = map(optimize_pair, pairs)`. The point is to get a tuple of concrete types rather than a vector of abstract types.

File: test/16_recode.jl (hunk 3 above)
Comment (pgagarinov) [reply test/16_recode.jl:12]: Ok but having just a benchmark provides weaker protection against an accidental performance degradation as opposed to performance test (because the benchmarks are not integrated into CI).

Comment (nalimilan) [reply test/16_recode.jl:12]: Yes but CI machines cannot reliably be used to check performance. They may have unexpected slowdowns, while BenchmarkTools uses many runs (normally on a dedicated machine) to avoid this kind of problem.

File: src/recode.jl [hunk 5]
```diff
@@ -49,16 +49,27 @@ A user defined type could override this method to define an appropriate test fun
...
 @inline recode_in(x, collection) = any(x ≅ y for y in collection)
 
+
+function optimize_pair(pair::Pair)::Pair
+    if typeof(pair.first) <: AbstractArray
+        pair = Set(pair.first) => pair.second
+    end
+    return pair;
+end
+
```
Comment (nalimilan) [inline src/recode.jl:59]:
```suggestion
optimize_pair(pair::Pair) = pair
optimize_pair(pair::Pair{<:AbstractArray}) = Set(pair.first) => pair.second
```

File: src/recode.jl [hunk 6]
```diff
@@ -49,16 +49,27 @@ A user defined type could override this method to define an appropriate test fun
...
 
+    opt_pairs = map(optimize_pair, pairs);
+
     @inbounds for i in eachindex(dest, src)
         x = src[i]
 
-        for j in 1:length(pairs)
-            p = pairs[j]
+        for j in 1:length(opt_pairs)
+            p = opt_pairs[j]
```
Comment (nalimilan) [inline src/recode.jl:72]:
```suggestion
        for j in 1:length(pairs)
            p = optimize_pair(pairs[j])
```

File: src/recode.jl [hunk 7]
```diff
@@ -49,16 +49,27 @@ A user defined type could override this method to define an appropriate test fun
...
+end
+
+
 function recode!(dest::AbstractArray{T}, src::AbstractArray, default::Any, pairs::Pair...) where {T}
     if length(dest) != length(src)
         throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
     end
 
+    opt_pairs = map(optimize_pair, pairs);
+
```
Comment (nalimilan) [inline src/recode.jl:67]:
```suggestion
```

File: src/recode.jl [hunk 8]
```diff
@@ -96,7 +107,9 @@ function recode!(dest::CategoricalArray{T}, src::AbstractArray, default::Any, pa
         throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
     end
 
-    vals = T[p.second for p in pairs]
+    opt_pairs = map(optimize_pair, pairs);
+
+    vals = T[p.second for p in opt_pairs]
```
Comment (nalimilan) [inline src/recode.jl:112]:
```suggestion
    vals = T[p.second for p in pairs]
```

File: src/recode.jl [hunk 9]
```diff
@@ -110,8 +123,8 @@ function recode!(dest::CategoricalArray{T}, src::AbstractArray, default::Any, pa
     @inbounds for i in eachindex(drefs, src)
         x = src[i]
 
-        for j in 1:length(pairs)
-            p = pairs[j]
+        for j in 1:length(opt_pairs)
+            p = opt_pairs[j]
```
Comment (nalimilan) [inline src/recode.jl:127]:
```suggestion
        for j in 1:length(pairs)
            p = optimize_pair(pairs[j])
```

File: benchmark/benchmarks.jl [hunk 10]
```diff
@@ -86,4 +86,16 @@ SUITE["repeated assignment"]["empty dest"] =
...
     @benchmarkable mycopy!(c2, a) setup = c2=copy(c)
 SUITE["repeated assignment"]["many levels dest"] =
-    @benchmarkable mycopy!(d2, a) setup = d2=copy(d)
\ No newline at end of file
+    @benchmarkable mycopy!(d2, a) setup = d2=copy(d)
+    
+function recode2none(orig_vec, cat2merge_vec)
+    recode(orig_vec, cat2merge_vec => "None");
+end
+
```
Comment (nalimilan) [inline benchmark/benchmarks.jl:94]:
```suggestion
```

File: benchmark/benchmarks.jl [hunk 11]
```diff
@@ -86,4 +86,16 @@ SUITE["repeated assignment"]["empty dest"] =
...
+orig_vec = (x -> repeat(x, 32)).(string.([x % 1000 for x in  1:1000000]));
+cat2merge_vec = (x -> repeat(x, 32)).(string.([x % 1000 for x in  1:100000]));
+SUITE["recode"] = BenchmarkGroup()
+SUITE["recode"]["vectors"] = @benchmarkable recode2none(orig_vec, cat2merge_vec);
+SUITE["recode"]["categorical_vectors"] = @benchmarkable recode2none(categorical(orig_vec), categorical(cat2merge_vec));
+SUITE["recode"]["matrices"] = @benchmarkable recode2none(reshape(orig_vec, :, 1), reshape(cat2merge_vec, :, 1));
+SUITE["recode"]["categorical_matrices"] = @benchmarkable recode2none(categorical(reshape(orig_vec, :, 1)), categorical(reshape(cat2merge_vec, :, 1)));
```
Comment (nalimilan) [inline benchmark/benchmarks.jl:101]:
```suggestion
SUITE["recode"]["vectors"] = @benchmarkable recode(orig_vec, cat2merge_vec => "None")
SUITE["recode"]["categorical_vectors"] = @benchmarkable recode(categorical(orig_vec), cat2merge_vec => "None")
SUITE["recode"]["matrices"] = @benchmarkable recode(reshape(orig_vec, :, 1), reshape(cat2merge_vec, :, 1) => "None")
SUITE["recode"]["categorical_matrices"] = @benchmarkable recode(categorical(reshape(orig_vec, :, 1)), reshape(cat2merge_vec, :, 1) => "None")
```

File: benchmark/benchmarks.jl [hunk 12]
```diff
@@ -86,4 +86,16 @@ SUITE["repeated assignment"]["empty dest"] =
...
-    @benchmarkable mycopy!(d2, a) setup = d2=copy(d)
\ No newline at end of file
+    @benchmarkable mycopy!(d2, a) setup = d2=copy(d)
+    
+function recode2none(orig_vec, cat2merge_vec)
+    recode(orig_vec, cat2merge_vec => "None");
+end
+
+orig_vec = (x -> repeat(x, 32)).(string.([x % 1000 for x in  1:1000000]));
+cat2merge_vec = (x -> repeat(x, 32)).(string.([x % 1000 for x in  1:100000]));
```
Comment (nalimilan) [inline benchmark/benchmarks.jl:96]:
```suggestion
orig_vec = (x -> repeat(x, 32)).(string.([x % 1000 for x in 1:1000000]))
cat2merge_vec = (x -> repeat(x, 32)).(string.([x % 1000 for x in 1:100000]))
```

**Comment by nalimilan** [comment]
Thanks!
